eth_p2p/rlpx_protocols/shh_protocol.nim

```diff
@@ -416,31 +539,406 @@ proc add*(self: var Queue, msg: Message) =
+  if not env.valid(): # actually just ttl !=0 is sufficient
+    return false
+
+  # We have to do the same checks here as in the messages proc not to leak
+  # any information that the message originates from this node.
+  let msg = initMessage(env)
+  if not msg.allowed(node.protocolState(shh).config):
+    return false
+
+  debug "Adding message to queue"
+  if node.protocolState(shh).queue.add(msg):
+    # Also notify our own filters of the message we are sending,
+    # e.g. msg from local Dapp to Dapp
+    node.protocolState(shh).filters.notify(msg)
+
+  return true
+
+proc postMessage*(node: EthereumNode, pubKey = none[PublicKey](),
+                  symKey = none[SymKey](), src = none[PrivateKey](),
+                  ttl: uint32, topic: Topic, payload: Bytes,
+                  padding = none[Bytes](), powTime = 1'f,
+                  targetPeer = none[NodeId]()): bool =
+  # NOTE: Allow a post without a key? Encryption is mandatory in v6?
+  let payload = encode(Payload(payload: payload, src: src, dst: pubKey,
+                               symKey: symKey, padding: padding))
+  if payload.isSome():
+    var env = Envelope(expiry:epochTime().uint32 + ttl + powTime.uint32,
+                       ttl: ttl, topic: topic, data: payload.get(), nonce: 0)
+
+    # Allow lightnode to post only direct p2p messages
+    if targetPeer.isSome():
+      return node.sendP2PMessage(targetPeer.get(), env)
+    elif not node.protocolState(shh).config.isLightNode:
+      # XXX: make this non blocking or not?
+      # In its current blocking state, it could be noticed by a peer that no
+      # messages are send for a while, and thus that mining PoW is done, and that
+      # next messages contains a message originated from this peer
+      env.nonce = env.minePow(powTime)
+      return node.sendMessage(env)
+    else:
+      error "Light node not allowed to post messages"
+      return false
+  else:
+    error "Encoding of payload failed"
+    return false
+
+proc subscribeFilter*(node: EthereumNode, filter: Filter,
+                      handler = none[FilterMsgHandler]()): string =
+  return node.protocolState(shh).filters.subscribeFilter(filter, handler)
+
+proc unsubscribeFilter*(node: EthereumNode, filterId: string): bool =
+  var filter: Filter
+  return node.protocolState(shh).filters.take(filterId, filter)
+
+proc getFilterMessages*(node: EthereumNode, filterId: string): seq[ReceivedMessage] =
+  return node.protocolState(shh).filters.getFilterMessages(filterId)
+
+proc filtersToBloom*(node: EthereumNode): Bloom =
+  return node.protocolState(shh).filters.toBloom()
+
+proc setPowRequirement*(node: EthereumNode, powReq: float64) {.async.} =
+  # NOTE: do we need a tolerance of old PoW for some time?
+  node.protocolState(shh).config.powRequirement = powReq
+  for peer in node.peers(shh):
+    # asyncCheck peer.powRequirement(cast[uint](powReq))
+    await peer.powRequirement(cast[uint](powReq))
```

Comment:
You can send these messages in parallel by first collecting all the futures and then calling `await all(...)`

Reply:
ah yes, I'll fix this